qiskit/transpiler/target.py

```diff
@@ -0,0 +1,387 @@
+        self.error = error
+        self.properties = properties
+
+
+class Target:
+    """
+    A description of gates on a backend. It exists around a central mapping of
+    :class:`~qiskit.circuit.Instruction` objects to their properties on the
+    device. As a basic example, lets assume your device is two qubits, supports
+    :class:`~qiskit.circuit.library.UGate` on both qubits and
+    :class:`~qiskit.circuit.library.CXGate` in both directions you would create
+    the gate map like::
+
+        from qiskit.transpiler import Target
+        from qiskit.circuit.library import UGate, CXGate
+        from qiskit.circuit import Parameter
+
+        gmap = Target()
+        theta = Parameter('theta')
+        phi = Parameter('phi')
+        lam = Parameter('lambda')
+        u_props = {
+            (0,): InstructionProperties(length=5.23e-8, error=0.00038115),
+            (1,): InstructionProperties(length=4.52e-8, error=0.00032115),
+        }
+        gmap.add_gate(UGate(theta, phi, lam), [(0,), (1,)], properties=u_props)
+        cx_props = {
+            (0,1): InstructionProperties(length=5.23e-7, error=0.00098115),
+            (1,0): InstructionProperties(length=4.52e-7, error=0.00132115),
+        }
+        gmap.add_gate(CXGate(), [(0, 1), (1, 0)], properties=cx_props)
+
+    .. note::
+
+        This class assumes that qubit indices start at 0 and are a contiguous
+        set if you want a submapping the bits will need to be reindexed in
+        a new object
+
+    .. note::
+
+        This class is designed to be additive only for gates, qargs, and qubits.
+        If you need to remove one of these the best option is to iterate over
+        an existing object and create a new subset (or use one of the methods
+        to do this). The object internally caches different views and these
+        would potentially be invalidated by removals.
+    """
+
+    __slots__ = (
+        "num_qubits",
+        "_gate_map",
+        "_gate_name_map",
+        "_qarg_gate_map",
+        "description",
+        "_coupling_graph",
+        "_unweighted_dist_matrix",
+        "_length_distance_matrix",
+        "_error_distance_matrix",
+        "_instruction_durations",
+    )
+
+    def __init__(self, description=None):
+        """
+        Create a new gate map class
+
+        Args:
+            gate_map (dict): A dictionary of gate_weight classes for keys and a list
+                qargs for
+            description (str): A string to describe the coupling map.
+        """
+        self.num_qubits = 0
+        self._gate_name_map = {}
+        self._gate_map = {}
+        self._qarg_gate_map = {}
+        self.description = description
+        self._coupling_graph = None
+        self._unweighted_dist_matrix = None
+        self._length_distance_matrix = None
+        self._error_distance_matrix = None
+        self._instruction_durations = None
+
+    def add_instruction(self, instruction, qargs, name=None, properties=None):
```

Comment:
Why `qargs` is necessary? It is already defined as the key of `properties`. Seems to me like this can cause some edge cases. I think qubit number validation is also missing.

Reply:
I did this mostly because properties aren't required. I was thinking of backends where there are no noise or durations reported (like aqt) and don't have any properties to set for an instruction. But looking back on this I agree it's redundant and error prone if `set(qargs) != set(properties)`. I'll update this and just take properties and document the case with no properties and just kwargs